providers/implementations/keymgmt/mlkem_kmgmt-c.inc

```diff
@@ -0,0 +1,596 @@
+
+struct mlkem_gen_ctx {
+    void *provctx;
+    uint8_t seed[MLKEM_SEED_BYTES];
+    int seeded;
+    int selection;
+};
+
+static void *mlkem_new(void *provctx)
+{
+    MLKEM_KEY *key = NULL;
+    OSSL_LIB_CTX *libctx = NULL;
+
+    if (!ossl_prov_is_running())
+        return 0;
+    if (provctx != NULL)
+        libctx = PROV_LIBCTX_OF(provctx);
+
+    key = OPENSSL_malloc(sizeof(MLKEM_KEY));
+    if (key != NULL) {
+        key->provctx = provctx;
+        key->mlkem_ctx = ossl_mlkem_newctx(libctx, NULL);
+        if (key->mlkem_ctx == NULL) {
+            OPENSSL_free(key);
+            return NULL;
+        }
+        key->keytype = 0;
+    }
+    return key;
+}
+
+static void mlkem_free(void *vkey)
+{
+    MLKEM_KEY *key = vkey;
+
+    if (key == NULL)
+        return;
+
+    /* Cleanse any private key material */
+    if (key->keytype == MLKEM_KEY_TYPE_PRIVATE)
+        cleanse(key);
+    ossl_mlkem_ctx_free(key->mlkem_ctx);
+    OPENSSL_free(key);
+}
+
+static int mlkem_has(const void *vkey, int selection)
+{
+    static int selectable =
+        OSSL_KEYMGMT_SELECT_PUBLIC_KEY |
+        OSSL_KEYMGMT_SELECT_PRIVATE_KEY;
+    const MLKEM_KEY *key = vkey;
+
+    if (!ossl_prov_is_running() || key == NULL)
+        return 0;
+    if ((selection & selectable) == 0)
+        return 1;
+
+    /*
+     * Fail if neither pub nor priv key is available.  The public key is
+     * available when either is provided.
+     */
+    if (key->keytype == 0
+        || ((selection & OSSL_KEYMGMT_SELECT_PRIVATE_KEY) != 0
+            && key->keytype != MLKEM_KEY_TYPE_PRIVATE)) {
+        ERR_raise(ERR_LIB_PROV, PROV_R_MISSING_KEY);
+        return 0;
+    }
+    return 1;
+}
+
+static int mlkem_match(const void *vkey1, const void *vkey2, int selection)
+{
+    static int selectable =
+        OSSL_KEYMGMT_SELECT_DOMAIN_PARAMETERS |
+        OSSL_KEYMGMT_SELECT_KEYPAIR;
+    const MLKEM_KEY *key1 = vkey1;
+    const MLKEM_KEY *key2 = vkey2;
+
+    if (!ossl_prov_is_running())
+        return 0;
+
+    /*-
+     * provider-keymgmt(7) documentation says:
+     *
+     *   It is assumed that the caller has ensured that *keydata1* and
+     *   *keydata2* are both owned by the implementation of this function.
+     *
+     * We take it this also implies both are non-NULL!
+     */
+    if ((selection & selectable) != 0)
+        if (key1->keytype != key2->keytype)
+            return 0;
+
+    /* We've checked the domain parameters */
+    if (!(selection & OSSL_KEYMGMT_SELECT_KEYPAIR))
+        return 1;
+
+    /* When neither have any key material, consider unequal */
+    if (key1->keytype == 0)
+        return 0;
+
+    return cmpkeys(key1, key2) == 0;
+}
+
+static int key_to_params(MLKEM_KEY *key, OSSL_PARAM_BLD *tmpl,
+                         OSSL_PARAM params[], int include_private)
+{
+    if (key == NULL)
+        return 0;
+
+    /* Nothing to do if we don't have any key material yet */
+    if (key->keytype == 0)
```

Comment:
We could add `|| key type > 2`, but I don't think it improves the code, it would have to be checked everywhere. Would probably need to change `keytype` to an enum, and hope that the compiler would complain about assignments outside the known valid values. The change to use an `enum` however may be worth it?